autopilot/scanner.go

```diff
@@ -2,89 +2,278 @@ package autopilot
+		s.performHostScans()
 		s.mu.Lock()
-		s.running = false
+		s.scanning = false
 		s.mu.Unlock()
 	}()
+	return true
 }
 
-func (s *scanner) performHostScan() {
-	// TODO: initial scan with much shorter
-	// scan up to 10 hosts that we haven't interacted with in at least 1 week
-	hosts, err := s.ap.bus.Hosts(time.Now().Add(-7*24*time.Hour), 10)
-	if err != nil {
-		return
-	}
-	type res struct {
-		hostKey  consensus.PublicKey
-		settings rhpv2.HostSettings
-		err      error
+func (s *scanner) tryUpdateTimeout() time.Duration {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if !s.isTimeoutUpdateRequired() {
+		return s.timeout
 	}
-	resChan := make(chan res)
-	for _, h := range hosts {
-		go func(h hostdb.Host) {
-			scan, err := s.ap.worker.RHPScan(h.PublicKey, h.NetAddress())
-			resChan <- res{h.PublicKey, scan.Settings, err}
-		}(h)
+
+	timeout, _ := s.tracker.timeout() // TODO: handle error
+	if timeout == 0 {
+		return 0
 	}
-	for range hosts {
-		var r res
 
+	s.timeoutLastUpdate = time.Now()
+	s.timeout = timeout
+	return s.timeout
+}
+
+// performHostScans scans every host in our database
+func (s *scanner) performHostScans() {
+	// TODO: running a scan on all hosts is not infinitely scalable, this will
+	// need to be updated to be smarter and run on a subset of all hosts
+	hosts, _ := s.bus.AllHosts()
+	s.pool.launchScans(hosts)
+	inflight := len(hosts)
+
+	// launch workers
+	respChan := s.pool.launchWorkers(func(resChan chan scanResp) {
+		for req := range s.pool.scanQueue {
+			if s.isStopped() {
+				return
+			}
+
+			timeout := s.tryUpdateTimeout()
+			scan, err := s.worker.RHPScan(req.hostKey, req.hostIP, timeout)
+			if scan.Ping > 0 {
+				s.tracker.addTiming(float64(time.Duration(scan.Ping).Milliseconds()))
+			}
+
+			resChan <- scanResp{req.hostKey, scan.Settings, err}
+		}
+	})
+
+	// defer a close of the response channel
+	defer func() {
+		// NOTE: we do not close the channel if the scan was interrupted and
+		// there are ongoing scans, this to avoid a race condition where a
+		// response is being sent down the channel after it's closed
+		if inflight == 0 {
+			close(respChan)
+		}
+	}()
+
+	// handle responses
+	for inflight > 0 {
+		var res scanResp
 		select {
-		case <-s.ap.stopChan:
+		case <-s.stopChan:
 			return
-		case r = <-resChan:
+		case res = <-respChan:
+			inflight--
 		}
 
-		if r.err != nil {
-			err := s.ap.bus.RecordHostInteraction(r.hostKey, hostdb.Interaction{
+		if res.err != nil {
+			err := s.bus.RecordHostInteraction(res.hostKey, hostdb.Interaction{
 				Timestamp: time.Now(),
-				Type:      "scan",
+				Type:      hostdb.InteractionScan,
 				Success:   false,
-				Result:    json.RawMessage(`{"error": "` + r.err.Error() + `"}`),
+				Result:    json.RawMessage(`{"error": "` + res.err.Error() + `"}`),
 			})
 			_ = err // TODO
 		} else {
-			js, _ := json.Marshal(r.settings)
-			err := s.ap.bus.RecordHostInteraction(r.hostKey, hostdb.Interaction{
+			js, _ := json.Marshal(res.settings)
```

Comment:
yes. I suppose it would be better to add a helper:
```go
func jsonMarshal(v interface{}) []byte {
    js, err := json.Marshal(v)
    if err != nil {
         panic(err)
    }
    return js
}
```
but marshalling never fails unless you try to encode a "weird" type, usually a `map` with an unsupported key type (like a `struct`), or a function or a channel.